File: java/src/main/java/com/meteroid/api/CheckoutSessions.java
// this file is @generated
package com.meteroid.api;

import com.meteroid.MeteroidHttpClient;
import com.meteroid.Utils;
import com.meteroid.exceptions.ApiException;
import com.meteroid.models.CancelCheckoutSessionResponse;
import com.meteroid.models.CreateCheckoutSessionRequest;
import com.meteroid.models.CreateCheckoutSessionResponse;
import com.meteroid.models.GetCheckoutSessionResponse;
import com.meteroid.models.ListCheckoutSessionsResponse;

import okhttp3.HttpUrl;

import java.io.IOException;

public class CheckoutSessions {
    private final MeteroidHttpClient client;

    public CheckoutSessions(MeteroidHttpClient client) {
        this.client = client;
    }

    /** */
    public ListCheckoutSessionsResponse listCheckoutSessions() throws IOException, ApiException {

        return this.listCheckoutSessions(new CheckoutSessionsListCheckoutSessionsOptions());
    }

    /** */
    public ListCheckoutSessionsResponse listCheckoutSessions(
            final CheckoutSessionsListCheckoutSessionsOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/checkout-sessions");
        if (options.customerId != null) {
            url.addQueryParameter("customer_id", Utils.serializeQueryParam(options.customerId));
        }
        if (options.status != null) {
            url.addQueryParameter("status", Utils.serializeQueryParam(options.status));
        }
        return this.client.executeRequest(
                "GET", url.build(), null, null, ListCheckoutSessionsResponse.class);
    }

    /** */
    public CreateCheckoutSessionResponse createCheckoutSession(
            final CreateCheckoutSessionRequest createCheckoutSessionRequest)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/checkout-sessions");
        return this.client.executeRequest(
                "POST",
                url.build(),
                null,
                createCheckoutSessionRequest,
                CreateCheckoutSessionResponse.class);
    }

    /** */
    public GetCheckoutSessionResponse getCheckoutSession(final String id)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/checkout-sessions/%s", id));
        return this.client.executeRequest(
                "GET", url.build(), null, null, GetCheckoutSessionResponse.class);
    }

    /** */
    public CancelCheckoutSessionResponse cancelCheckoutSession(final String id)
            throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/checkout-sessions/%s/cancel", id));
        return this.client.executeRequest(
                "POST", url.build(), null, null, CancelCheckoutSessionResponse.class);
    }
}

File: go/checkout_sessions.go
// this file is @generated
package meteroid

import (
	"context"
	"net/http"
)

// CheckoutSessionsListCheckoutSessionsOptions carries the query and header parameters of
// CheckoutSessions.ListCheckoutSessions.
//
// Optional parameters are pointers; leave them nil to omit them.
type CheckoutSessionsListCheckoutSessionsOptions struct {
	CustomerId *CustomerId

	Status *CheckoutSessionStatus
}

// CheckoutSessions groups the checkout sessions operations of the Meteroid API.
type CheckoutSessions struct {
	client *Client
}

func (a *CheckoutSessions) ListCheckoutSessions(ctx context.Context, options *CheckoutSessionsListCheckoutSessionsOptions) (*ListCheckoutSessionsResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/checkout-sessions")

	if options != nil {
		if options.CustomerId != nil {
			req.SetQueryParam("customer_id", string(*options.CustomerId))
		}
		if options.Status != nil {
			req.SetQueryParam("status", string(*options.Status))
		}
	}

	var out ListCheckoutSessionsResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *CheckoutSessions) CreateCheckoutSession(ctx context.Context, createCheckoutSessionRequest CreateCheckoutSessionRequest) (*CreateCheckoutSessionResponse, error) {
	req := newRequest(http.MethodPost, "/api/v1/checkout-sessions")

	req.SetJSONBody(createCheckoutSessionRequest)

	var out CreateCheckoutSessionResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *CheckoutSessions) GetCheckoutSession(ctx context.Context, id string) (*GetCheckoutSessionResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/checkout-sessions/{id}")
	req.SetPathParam("id", id)

	var out GetCheckoutSessionResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *CheckoutSessions) CancelCheckoutSession(ctx context.Context, id string) (*CancelCheckoutSessionResponse, error) {
	req := newRequest(http.MethodPost, "/api/v1/checkout-sessions/{id}/cancel")
	req.SetPathParam("id", id)

	var out CancelCheckoutSessionResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
